Add variant support for slugs & server version limitation

Pablete1234 opened the pull request:
```xml
<map min-server-version="1.21">
    <name>New map</name>
    <variant id="old" min-server-version="1.8" max-server-version="1.17" world="legacy">Old map</variant>
</map>
```

The map itself can have `min-server-version` and `max-server-version` and individual variants can also override this

# Slugs
Currently slugs are broken in variants, as one variant can't know the slug of other variants, since it's not part of the XML it sees (part of conditionals)
This PR fixes that by adding slug as an attribute in variants:

```xml
<map>
  <name>New map</name>
  <variant id="old" slug="map_old">Old map</variant>
</map>
```
If a slug is defined in the variant, that slug is used, otherwise:
For variants `override=true`: the slug defaults to the variant name slugified
For variants `override=false` & no default `<slug>`: the slug will be the full variant name slugified
For variants `override=false` & and default `<slug>`: the slug will be `defaultslug + "_" + variantId`

This aligns closely with what the behavior would usually be if it went thru the old rules, but won't always be the same

Pablete1234 committed:
Add support for version-gated maps & variants

Signed-off-by: Pablo Herrera <pabloherrerapalacio@gmail.com>

TTtie commented on core/src/main/java/tc/oc/pgm/map/MapInfoImpl.java:
```diff
@@ -367,10 +370,13 @@ private static class VariantData implements VariantInfo {
     public VariantData(Element root, @Nullable Element variantEl) throws InvalidXMLException {
       String name = assertNotNull(Node.fromRequiredChildOrAttr(root, "name").getValueNormalize());
       String slug = assertNotNull(root).getChildTextNormalize("slug");
+      Node minVer = Node.fromAttr(root, "min-server-version");
```
Modern (1.9+) worlds have got version metadata in `level.dat`, maybe the minimum server version could be inferred from there if not specified?

Pablete1234 replied:
Having this in the XML makes it more visible to mapdevs, and the behavior more predictable. Additionally, there's times where a map may be made in 1.9 or 1.12 and yet be loaded and play in 1.8 just fine. Forcibly making those not load would be a behavior change.

I'm fine with having a default min-server-version of 1.13 if the world is in 1.13+ format, as that is never going to be able to load in 1.8, but otherwise keep it xml defined

calcastor replied:
> I'm fine with having a default min-server-version of 1.13 if the world is in 1.13+ format, as that is never going to be able to load in 1.8, but otherwise keep it xml defined

This seems ideal and aligns similarly with how PGM tends to treat `level.dat` information at present.

TTtie committed:
fix: avoid loading 1.13+ maps in legacy versions

This patch adds the ability to prevent 1.13+ maps without a specified minimum version in `map.xml` from accidentally being loaded into legacy servers, causing undefined behavior including server crashes when cycled to.

This is achieved by reading the data version (an integer increased with any Minecraft version, including snapshots [1]) from `level.dat` bundled with the map and inferring whether the map is a 1.13+ map.

[1]: https://minecraft.wiki/w/Data_version

Signed-off-by: TTtie <me@tttie.cz>

Pablete1234 committed:
Minor cleanup to level.dat reading changes

Signed-off-by: Pablo Herrera <pabloherrerapalacio@gmail.com>

TTtie commented on platform/platform-sportpaper/src/main/java/tc/oc/pgm/platform/sportpaper/SpMiscUtil.java:
```diff
@@ -83,4 +89,19 @@ public ThrownPotion spawnPotion(Location loc, ItemStack item) {
+      return dataTag.hasKeyOfType("DataVersion", NBT_TAG_ANY_NUMERIC)
+          ? dataTag.getInt("DataVersion")
+          : -1;
+    } catch (Exception ignored) {
```
style: might be worthwhile to catch all throwables to keep it consistent with modern
```suggestion
    } catch (Throwable ignored) {
```

TTtie commented on platform/platform-sportpaper/src/main/java/tc/oc/pgm/platform/sportpaper/SpMiscUtil.java:
```diff
@@ -83,4 +89,19 @@ public ThrownPotion spawnPotion(Location loc, ItemStack item) {
+  public int getWorldDataVersion(Path levelDatPath) {
+    try {
+      var compoundTag =
+          NBTCompressedStreamTools.a(Files.newInputStream(levelDatPath, StandardOpenOption.READ));
```
style: keep the root tag name consistent with modern
```suggestion
      var root =
          NBTCompressedStreamTools.a(Files.newInputStream(levelDatPath, StandardOpenOption.READ));
```

TTtie commented on core/src/main/java/tc/oc/pgm/map/MapInfoImpl.java:
```diff
@@ -384,9 +393,23 @@ public VariantData(Element root, @Nullable Element variantEl) throws InvalidXMLE
+        if (minVerVariant != null) minVer = minVerVariant;
+
+        Node maxVerVariant = Node.fromAttr(variantEl, "max-server-version");
+        if (maxVerVariant != null) minVer = minVerVariant;
```
🚨💥 variant-level max-server-version setting is being straight up ignored
```suggestion
        if (maxVerVariant != null) maxVer = maxVerVariant;
```

testing diff over OCC PublicMaps:

```diff
diff --git a/ctf/directrix/map.xml b/ctf/directrix/map.xml
index 9811fc9e..525771c0 100644
--- a/ctf/directrix/map.xml
+++ b/ctf/directrix/map.xml
@@ -3,8 +3,8 @@
 <variant id="halloween" world="halloween" override="true">DirecTricks</variant>
 <variant id="christmas" world="christmas" override="true">DirectriXmas</variant>
 <variant id="classic" world="classic">Classic</variant>
-<variant id="king_ranch" world="king_ranch" override="true">King Ranch</variant>
-<variant id="king_ranch_og" world="king_ranch_og" override="true">King Ranch: OG</variant>
+<variant id="king_ranch" world="king_ranch" override="true" min-server-version="1.0" max-server-version="1.8.8">King Ranch</variant>
+<variant id="king_ranch_og" world="king_ranch_og" override="true" min-server-version="1.21">King Ranch: OG</variant>
 <unless variant="king_ranch,king_ranch_og">
     <version>1.1.5</version>
 </unless>
```

TTtie reviewed (changes requested):
uh-oh! it blew up :sob:

add'l issues/notes:
- `/map <map>` lists variants not loaded due to version restrictions 
- it might be worthwhile to note somewhere in the docs that `max-server-version="1.8"` means "limit to 1.8.0", not "limit to 1.8.8" (depending on how common the belief of "`max-server-version="1.8"` includes `1.8.8` servers" is going to get)

Pablete1234 committed:
Improve variant support & fix PR comments

Signed-off-by: Pablo Herrera <pabloherrerapalacio@gmail.com>

Pablete1234 commented:
![image](https://github.com/user-attachments/assets/7ee2d933-c924-4992-a5ff-f037c08533a3)
![image](https://github.com/user-attachments/assets/e7d22364-ec00-4713-b72f-a7eeeb3b5584)

Variants that can't load will now appear in red and have an explicit hover message

Pablete1234 committed:
Variants with override will also override slug

Signed-off-by: Pablo Herrera <pabloherrerapalacio@gmail.com>

TTtie commented on core/src/main/java/tc/oc/pgm/map/MapInfoImpl.java:
```diff
@@ -349,44 +356,64 @@ protected void setContext(MapContextImpl context) {
+        else if (override) slug = null;
+        else if (slug != null) slug += "_" + variantId;
+
+        minVer = fallback(Node.fromAttr(variantEl, "min-server-version"), minVer);
+        maxVer = fallback(Node.fromAttr(variantEl, "max-server-version"), maxVer);
```
~~supposed to be other way around? (otherwise root-level map settings will take priority over variant-level settings)~~ not the issue
```suggestion
        minVer = fallback(minVer, Node.fromAttr(variantEl, "min-server-version"));
        maxVer = fallback(maxVer, Node.fromAttr(variantEl, "max-server-version"));
```

Pablete1234 replied:
Nope, it takes the first non null param

This is just "take the variant's node, otherwise fallback to the map's node"

TTtie replied:
~~Currently, if both the default variant and a variant define a version bound (such as the example in the cover letter), the first non-null parameter in that case the root-level `{min,max}-server-version`, i. e. the `{min,max}-server-version` in `<map>` would take precedence over the attributes in `<variant>`.~~

~~In other words, root-level/default variant version bound is used first, falling back to the variant-level bound if not present 🤷‍♀️~~ 

Wait, nevermind.

TTtie replied:
This was observed and read incorrectly on my side, sorry! (╯°□°)╯︵ ┻━┻


There's actually a different issue with this – the root-level version bounds get unused node warnings when a variant overrides them… although that's correct, that should get silenced as it is intended to not be used

Pablete1234 commented:
There's one pending thing and it's unused node detection can mark the attributes on the root as unused for the variants. How exactly we tackle unused node detection in variants is something i want to give more thought and a more final solution later down the road, so it'll be ignored for now.

cswhite2000 commented on core/src/main/java/tc/oc/pgm/map/MapLibraryImpl.java:
```diff
@@ -226,16 +223,22 @@ private MapContext loadMap(MapSource source, @Nullable String mapId) throws MapE
-        info.getId(), info, (m1, m2) -> m2.getVersion().isOlderThan(m1.getVersion()) ? m1 : m2);
+    // Only if from a supported version, add it to our library
+    if (info.isServerSupported()) {
+      maps.merge(
```
Will this register as a failure and should it? I feel like there would be some value in logging it specifically as a version incompatibility. I feel like people might be confused if we don't log anything.

Pablete1234 replied:
It will simply skip maps that are not for the current server version. It is intended as you may maintain a repo that contains both 1.8 and 1.21 maps, and simply when running in 1.8 the 1.21 maps are skipped and all is good

I'm not sure about logging as it'd likely lead to log-spam that is not really serving any use, as in, whoever set those maps knows what they're doing

cswhite2000 reviewed (commented):
No major concerns
